patches/server/0739-Implemented-OP-level-API.patch

```diff
@@ -0,0 +1,252 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: MCMDEV <john-m.1@gmx.de>
+Date: Thu, 5 Aug 2021 17:00:10 +0200
+Subject: [PATCH] Implemented OP level API
+
+
+diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+index 918f5221e94cbc867349c69c83563e225d2fef1d..afc9c3f0550053126640f570ca4a31b5e3584459 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+@@ -52,6 +52,17 @@ public class DedicatedPlayerList extends PlayerList {
+         this.saveOps();
+     }
+ 
++    //Paper start - OP level API
++    public int getOperatorLevel(GameProfile profile)    {
++        return super.getOperatorLevel(profile);
++    }
++
++    public void setOperatorLevel(GameProfile profile, int value)    {
++        super.setOperatorLevel(profile, value);
++        this.saveOps();
++    }
++    //Paper end
++
+     @Override
+     public void reloadWhiteList() {
+         this.loadWhiteList();
+diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
+index 48045993c8ad4b014cf4a67f7c4db42e014d1c81..7a6d9a037bcc47f19768e6a716a1aa918d82aeb1 100644
+--- a/src/main/java/net/minecraft/server/players/PlayerList.java
++++ b/src/main/java/net/minecraft/server/players/PlayerList.java
+@@ -1144,6 +1144,18 @@ public abstract class PlayerList {
+ 
+     }
+ 
++    //Paper start - OP level API
++    public int getOperatorLevel(GameProfile profile)   {
++        return this.getServer().getProfilePermissions(profile);
++    }
++
++    public void setOperatorLevel(GameProfile profile, int permissionLevel)  {
```

Comment:
Instead of completely re-implementing the op/deop logic defined in both PlayerList and DedicatedPlayerList couldn't this use a "deob-helper" style implementation along the lines of:

```java
public void op(GameProfile profile) { // Old method signature
  this.op(profile, this.server.getOperatorUserPermissionLevel())
}
public void op(GameProfile profile, int level) {
  // Prior old logic block, now using level instead of "hardcoded" server default level
}
```

Concerning that the ServerOpList will discard prior op entries if you call op again, re-opping a player with a new op level should pose no issue whatsoever.